### PR TITLE
Added QWERTY Euclidean distance (qwerty_levenshtein)

### PR DESCRIPTION
**Pull request by RJ-lovering**
Also added backtrace and and reader-friendly operations summary for qwerty_levenshtein.

### CODE REVIEW

**Comment by jamesturk** [comment]
I can't find any reference to this algorithm via Google, mind providing a source?

Also it looks like this is failing tests, but we can wait to discuss those until a decision has been made about inclusion

**Comment by RJ-lovering** [comment]
I'm not sure of its independent algorithmic name, myself, or if it has one.
It's just a variation on levenshtein-damerau that calculates substitution
cost sensitive to the physical distance between keys on a a QWERTY keyboard
and the effort of pressing SHIFT (the z value of the coordinate system). I
got the idea from this stack overflow exchange
<http://stackoverflow.com/questions/29233888/edit-distance-such-as-levenshtein-taking-into-account-proximity-on-keyboard>,
and developed it a bit further to accommodate some of the data I was
working with, in pursuit of finding duplicate entries for names of people
and medications in a database I work with.

It has served reasonably well as a component in glomming together typos
without bias toward higher-scoring in cases where letters at the beginning
of a word are the same (I did not find any research to support the idea
that typos are significantly more likely to occur at the end of a word than
at the beginning, and it was causing some tedium in the manual review phase
of my own project to weight them differently).

If it's not something that jellyfish ought to have, I'm happy to keep it on
my branch and keep that branch to myself, but I thought others might find
it useful.

**Comment by jamesturk** [comment]
I think it'd be tricky to keep up the maintenance on my end (and in keeping
parity w/ the others I'd need to port it to C & Go) - it is a neat idea
though, have you thought of rolling your own package?

**Comment by RJ-lovering** [comment]
I hadn't until you made the suggestion, actually, but since this is one in
a suite of tools I'm developing for my department, I might make a go of it
once I've figured out what's got the strongest reuse value. Thanks for the
suggestion!

I will, in the meantime, keep my branch to myself. Jellyfish is great work,
thanks for sharing it and keeping it up.

**Comment by jamesturk** [comment]
sounds good & thanks for understanding :)

File: jellyfish/_jellyfish.py [hunk 1]
```diff
@@ -111,6 +113,133 @@ def _jaro_winkler(ying, yang, long_tolerance, winklerize):
...
+                    new_target=re.sub(cause,'',target)
+                    return qwerty_levenshtein(new_target, source)
+            
+            insert = distance[i-1][j]+1
+            delete = distance[i][j-1]+1
+            substitute = distance[i-1][j-1]+sub_cost
+            
+            distance[i][j] = min(insert,delete,substitute)
+
+            if i and j and target[i]==source[j-1] and target[i-1]==source[j]:
```
Comment (brucechou1983) [inline jellyfish/_jellyfish.py]: IndexError here, should be

`if i and j and target[i-1] == source[j-2] and target[i-2] == source[i-1]`

**Pull request closed without merging**